datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java

```diff
@@ -0,0 +1,122 @@
+    _throughputProvider = throughputProvider;
+    _sourceClusterResolver = sourceClusterResolver;
+  }
+
+  @Override
+  public Map<String, Set<DatastreamTask>> assignPartitions(Map<String, Set<DatastreamTask>> currentAssignment,
+      DatastreamGroupPartitionsMetadata datastreamPartitions) {
+    DatastreamGroup datastreamGroup = datastreamPartitions.getDatastreamGroup();
+    String datastreamGroupName = datastreamGroup.getName();
+    HashMap<String, ClusterThroughputInfo> partitionThroughputInfo;
+
+    // Attempting to retrieve partition throughput info with a fallback mechanism to StickyPartitionAssignmentStrategy
+    try {
+      partitionThroughputInfo = fetchPartitionThroughputInfo();
+    } catch (RetriesExhaustedException ex) {
+      LOG.warn("Attempts to fetch partition throughput timed out. Falling back to regular partition assignment strategy");
+      return super.assignPartitions(currentAssignment, datastreamPartitions);
+    }
+
+    LOG.info("Old partition assignment info, assignment: {}", currentAssignment);
+    Validate.isTrue(currentAssignment.size() > 0,
+        "Zero tasks assigned. Retry leader partition assignment.");
+
+    // Calculating task count and assigned partitions
+    int taskCount = getTaskCountForDatastreamGroup(currentAssignment, datastreamGroupName);
+    List<String> assignedPartitions = getAssignedPartitionsForDatastreamGroup(currentAssignment, datastreamGroupName);
+
+    // Elastic task count validation
+    if (getEnableElasticTaskAssignment(datastreamGroup)) {
+      if (assignedPartitions.isEmpty()) {
+        performElasticTaskCountValidation(datastreamPartitions, taskCount);
+      }
+      updateOrRegisterElasticTaskAssignmentMetrics(datastreamPartitions, taskCount);
+    }
+
+    // Calculating unassigned partitions
+    List<String> unassignedPartitions = new ArrayList<>(datastreamPartitions.getPartitions());
+    unassignedPartitions.removeAll(assignedPartitions);
+
+    // Getting a task count estimate based on throughput
+    String clusterName = _sourceClusterResolver.getSourceCluster(datastreamPartitions.getDatastreamGroup());
+    LoadBasedTaskCountEstimator taskCountEstimator = new LoadBasedTaskCountEstimator();
+    ClusterThroughputInfo clusterThroughputInfo = partitionThroughputInfo.get(clusterName);
+    int taskCount2 = taskCountEstimator.getTaskCount(clusterThroughputInfo, assignedPartitions,
+        unassignedPartitions);
+
+    // Picking the maximum task count between elastic task count and throughput based task count estimate
+    int maxTaskCount = Math.max(taskCount, taskCount2);
+
+    // Doing assignment
+    LoadBasedPartitionAssigner partitionAssigner = new LoadBasedPartitionAssigner();
+    return partitionAssigner.assignPartitions(clusterThroughputInfo, assignedPartitions, unassignedPartitions,
+        maxTaskCount);
+  }
+
+  private HashMap<String, ClusterThroughputInfo> fetchPartitionThroughputInfo() {
+    PollUtils.poll(() -> {
```

Comment:
The poll method is blocking, but it will return something before the specified timeout. I configured the call to throw an exception if the retries are exhausted.